Detsim get psf functions

gondiaz opened the pull request:
Following issue #691:
10. upload functions to read an manipulate kr maps and sipm psfs for their transformation into a callable that can be used in `pes_at_sensors` function from PR #697 in module invisible_cities/cities/detsim_get_psf.py

This module contains two functions to get the referred psfs (pmt and sipm), an auxiliary function to construct a required callable to be used at `pes_at_sensors` and a psf temporal analytical function.

The psf analytical function must be substituted in the future by a light table equivalent.

gondiaz committed:
Add functions to create psf callable

gondiaz committed:
Correct bug in xy_function

gondiaz committed:
Add tests and sipm psf test data

andLaing commented on invisible_cities/cities/detsim_get_psf.py:
```diff
@@ -0,0 +1,66 @@
+import tables as tb
+import pandas as pd
+
+from typing import Tuple, Callable
```
Line per import

andLaing commented on invisible_cities/cities/detsim_get_psf.py:
```diff
@@ -0,0 +1,66 @@
+        y = np.clip(y, ymin, ymax-0.1*dy)
+        ix = ((x-xmin)//dx).astype(int)
+        iy = ((y-ymin)//dy).astype(int)
+        return array[ix, iy]
```
I'd prefer it had a more explicative name, function seems a bit generic as a name.

Also, you don't use z. Did you forget to push part of the code or is z not used? It shouldn't be an argument of the function if it has no use.

Also the type hint for a numpy array is np.ndarray. I'd give it a descriptive name other than just array which is confusing.

gondiaz replied:
Now, I changed this `create_xy_function` by `create_xyx_function`.  The fact of adding the `z` variable is because when we compute `pes = photons * psf(dxs, dys, dzs)` (PR #697 in `pes_at_sensors ` function), this `psf` is in general a x, y, z function. 

For krmap and sipmpsf, z dimension is not used, but it is added to do not include something duplicate like:
```
if S2:
    pes = photons * psf(dxs, dys, dzs)
elif S1:
    pes = photons * psf(dxs, dys)
```
Besides, we can add light tables, which are x, y, z functions.

andLaing commented on invisible_cities/cities/detsim_get_psf_test.py:
```diff
@@ -0,0 +1,88 @@
+    xaxis = (xmin, xmax, dx)
+    yaxis = (ymin, ymax, dy)
+
+    func = lambda x, y: - x**2
```
why does it have y as an argument if it only uses x?

gondiaz replied:
No reason for that, the function `func` is just a x, y, z (new code adds the z) for testing `create_xyz_function`. In the case it is there, just a x dependence is assumed.

andLaing replied:
It's a lambda function in a fixture that takes 3 arguments and only uses one?! Why? If you want it as a place holder for a more complex function which will use the other arguments write the full function and make it return `-x**2` if y and z are 0 or something like that. If it's always just going to give `-x**2` than just define it with one parameter `x`

gondiaz replied:
This is just a matter of taste. This function can be modified to whatever xyz function you want. Note that its aim is in `test_create_xyz_function`. I changed it with explicit x, y, z dependence inside, maybe now it looks better for you.

andLaing replied:
Well no, but thanks for changing it so that y and z are relevant

andLaing commented on invisible_cities/cities/detsim_get_psf_test.py:
```diff
@@ -0,0 +1,88 @@
+
+    #test create_function
+    H, _ = np.histogramdd((x, y), weights=f, bins=(xbins, ybins))
+    callable = create_xy_function(H, xaxis, yaxis)
```
give it a name, callable is a type

andLaing reviewed (changes requested):
Needs to be moved too.

Few clarifications needed on the individual functions

gondiaz committed:
Add LT functions and binedges_from_bincenters

gondiaz committed:
Remove redundancy

It is not necessary to have create_xy_function and
create_xyz_function. The xyz function is enough. Functions that
use xy, will be modified to include a thrid z dummy dimension.

gondiaz committed:
Modify psf functions be used as xyz function

gondiaz committed:
Remove detached axis to bins in get_ligthtables

gondiaz committed:
Update test to account for new create_xyz_function

gondiaz committed:
Line per import

gondiaz committed:
Change name callable to fxyz

gondiaz committed:
Change function to test cretate_xyz_function

gondiaz committed:
Change len by shape and nan by 0

gondiaz committed:
New detsim psf module

gondiaz committed:
Remove test module

gondiaz committed:
New psf module from loop_based

gondiaz committed:
Correct typo

gondiaz committed:
Remove unused import

gondiaz committed:
Add test file

gondiaz committed:
Cosmetics

gondiaz committed:
Add and modify Exceptions

gondiaz committed:
Add test for binedges_from_bincenters

gondiaz committed:
Clean commented code

gondiaz committed:
Add units in get_psf keyword arguments

gondiaz committed:
Modify get_psf to create_effective_psf

The function is renamed to be explanatory. The new output contains
n_time_bins and s2_sipm_nsamples. n_time_bins contains the effective
number of EL partitions (probably find a better name). s2_sipm_nsamples the number of samples in
the final waveform in which pes are distributed

gondiaz committed:
Add psf fixture and test function

gondiaz committed:
Remove unecessary wrapper function

gondiaz committed:
Add lighttables fixture and test

gondiaz committed:
Add Lighttables and PSF test-data for Detsim
